File: frontend/src/utils/arithmeticVerifier.ts
import type { ArithmeticWarning, FinancialSummary, LineItem } from '../types/supplement'

const TOLERANCE = 0.01

export function verifyArithmetic(
  lineItems: LineItem[],
  financialSummary: FinancialSummary,
): ArithmeticWarning[] {
  const warnings: ArithmeticWarning[] = []

  if (lineItems.length > 0) {
    const computed = lineItems.reduce((sum, item) => sum + item.claimed_amount, 0)
    if (Math.abs(computed - financialSummary.claimed_total) > TOLERANCE) {
      warnings.push({
        position: 'FINANCIAL_SUMMARY',
        claimed_amount: financialSummary.claimed_total,
        expected_amount: Math.round(computed * 100) / 100,
        difference: Math.round((financialSummary.claimed_total - computed) * 100) / 100,
      })
    }
  }

  const breakdown =
    financialSummary.disputed_amount +
    financialSummary.approvable_amount +
    financialSummary.conditional_amount
  if (Math.abs(breakdown - financialSummary.claimed_total) > TOLERANCE) {
    warnings.push({
      position: 'FINANCIAL_BREAKDOWN',
      claimed_amount: financialSummary.claimed_total,
      expected_amount: Math.round(breakdown * 100) / 100,
      difference: Math.round((financialSummary.claimed_total - breakdown) * 100) / 100,
    })
  }

  return warnings
}

export function formatCHF(amount: number): string {
  return `CHF ${amount.toLocaleString('de-CH', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`
}

export function formatPct(value: number): string {
  return `${value.toFixed(1)}%`
}

File: backend/services/arithmetic_verifier.py
from backend.models.schemas import SupplementReport, ArithmeticWarning

TOLERANCE = 0.01


def verify_arithmetic(report: SupplementReport) -> list[ArithmeticWarning]:
    """
    Verify financial arithmetic in the supplement report.
    Returns list of ArithmeticWarning for any mismatches found.
    """
    warnings: list[ArithmeticWarning] = []

    fs = report.financial_summary
    line_items = report.line_items

    # Check: sum of all line item claimed_amounts equals claimed_total
    if line_items:
        computed_total = sum(item.claimed_amount for item in line_items)
        if abs(computed_total - fs.claimed_total) > TOLERANCE:
            warnings.append(
                ArithmeticWarning(
                    position="FINANCIAL_SUMMARY",
                    claimed_amount=fs.claimed_total,
                    expected_amount=round(computed_total, 2),
                    difference=round(fs.claimed_total - computed_total, 2),
                )
            )

    # Check: disputed + approvable + conditional = claimed_total
    computed_breakdown = (
        fs.disputed_amount + fs.approvable_amount + fs.conditional_amount
    )
    if abs(computed_breakdown - fs.claimed_total) > TOLERANCE:
        warnings.append(
            ArithmeticWarning(
                position="FINANCIAL_BREAKDOWN",
                claimed_amount=fs.claimed_total,
                expected_amount=round(computed_breakdown, 2),
                difference=round(fs.claimed_total - computed_breakdown, 2),
            )
        )

    # Check: net_payable = approvable_amount - retention_applicable
    expected_net = fs.approvable_amount - fs.retention_applicable
    if abs(expected_net - fs.net_payable) > TOLERANCE:
        warnings.append(
            ArithmeticWarning(
                position="NET_PAYABLE",
                claimed_amount=fs.net_payable,
                expected_amount=round(expected_net, 2),
                difference=round(fs.net_payable - expected_net, 2),
            )
        )

    return warnings
